src/txmempool.cpp

```diff
@@ -57,46 +57,75 @@ size_t CTxMemPoolEntry::GetTxSize() const
+    // set up the new_cache_line to contain all of our transaction's children (note --
+    // already de-duplicated in case multiple outputs of ours are spent in one
+    // transaction).
+    // Always creates a new cache line entry.
+    vecEntries& new_cache_line = cache[update_it];
+    assert(new_cache_line.empty());
+    // next_idx_to_walk index keeps track of the elements that we've
+    // already expanded.
+    // Invariants:
+    // If index is < next_idx_to_walk, we've already walked it.
+    // If index is >= next_idx_to_walk, we need to walk it.
+    // If next_idx_to_walk >= new_cache_line.size(), we're finished.
+    size_t next_idx_to_walk = 0;
+    txiter next_it = update_it;
+    do {
+        for (const txiter child_it : GetMemPoolChildren(next_it)) {
+            // N.B. children can also be grand children, so guarding
+            // here is neccessary
+            if (visited(child_it)) continue;
+            // if it exists in the cache, copy cached descendants
+            // - All elements in cache are not excluded, no need to check when adding.
+            // - All keys in the cache are already excluded, no need to add
+            //   `child_it` to the new_cache_line (but not all excluded keys
+            //   are in the cache yet!)
+            cacheMap::iterator cached_grand_children = cache.find(child_it);
+            if (cached_grand_children != cache.end()) {
+                for (const txiter grand_child_it : cached_grand_children->second) {
+                    if (visited(grand_child_it)) continue;
+                    // place on the back and then swap into the next_idx_to_walk index
+                    // so we don't walk it ourselves (whoever put the grand
+                    // child in the cache must have already traversed this, so it must
+                    // not be excluded)
+                    new_cache_line.emplace_back(grand_child_it);
+                    std::swap(new_cache_line[next_idx_to_walk], new_cache_line.back());
+                    ++next_idx_to_walk;
                 }
-            } else if (!setAllDescendants.count(childEntry)) {
+            } else {
                 // Schedule for later processing
-                stageEntries.insert(childEntry);
+                // Element must not be excluded, or it would be in the cache already.
+                new_cache_line.emplace_back(child_it);
             }
         }
+
+        // finish loop here!
+        if (next_idx_to_walk >= new_cache_line.size()) break;
+
+        // Prepare for next iteration:
+        next_it = new_cache_line[next_idx_to_walk];
+        ++next_idx_to_walk;
+    } while (true);
+
+    // new_cache_line now contains all in-mempool descendants of update_it,
+    // compute updates now.
+    int64_t modify_size = 0;
+    CAmount modify_fee = 0;
+    int64_t modify_count = 0;
+    // All entries in the new_cache_line have
+    // already been checked against excluded list
```

Comment:
I'm confused with "... have already been checked against _excluded list_". Which list?

Reply:
mapMemPoolDescebdabtsToUpdate grows to contain the set of things to exclude while processing.

Previously this list was explicitly constructed, now it's implicit

Reply:
> Previously this list was explicitly constructed, now it's implicit

Yes. And the new logic does not apply any rule or list "to exclude" an element. It seems, the comments should avoid any "excluding" notions, and describe the logic in terms of "including" or "conditional including".

Reply:
Well the list does exist -- it's vHashesToUpdate. We just don't need to actually check or construct it because it's an implicit invariant held by the code.

I can't comment on the semantic difference between excluding or conditional including, but it does seem that what we're doing is excluding.

Reply:
In any case this comment & the function comment serve to document what the expected invariant is for the caller, so I prefer to leave this comment.